src/compass/utils/cloud_optimized_geotiff.py

```diff
@@ -0,0 +1,498 @@
+                            block_trailer_last_4_bytes_repeated,
+                            mask_interleaved_with_imagery)
+            if ovr_band.GetMaskFlags() == gdal.GMF_PER_DATASET and \
+                    (filename + '.msk') not in ds.GetFileList():
+                full_check_band(f, 'Mask band of overview %d' % i,
+                                ovr_band.GetMaskBand(), errors,
+                                block_order_row_major,
+                                block_leader_size_as_uint4,
+                                block_trailer_last_4_bytes_repeated, False)
+        gdal.VSIFCloseL(f)
+
+    return warnings, errors, details
+
+
+def validate(argv):
+    """Return 0 in case of success, 1 for failure."""
+
+    i = 1
+    filename = None
+    quiet = False
+    full_check = None
+    while i < len(argv):
+        if argv[i] == '-q':
+            quiet = True
+        elif argv[i] == '--full-check=yes':
+            full_check = True
+        elif argv[i] == '--full-check=no':
+            full_check = False
+        elif argv[i] == '--full-check=auto':
+            full_check = None
+        elif argv[i][0] == '-':
+            return Usage()
+        elif filename is None:
+            filename = argv[i]
+        else:
+            return Usage()
+
+        i += 1
+
+    if filename is None:
+        return Usage()
+
+    if full_check is None:
+        full_check = filename.startswith('/vsimem/') or os.path.exists(filename)
+
+    try:
+        ret = 0
+        warnings, errors, details = validate_cog(filename,
+                                                 full_check=full_check)
+        if warnings:
+            if not quiet:
+                print('The following warnings were found:')
+                for warning in warnings:
+                    print(' - ' + warning)
+                print('')
+        if errors:
+            if not quiet:
+                print('%s is NOT a valid cloud optimized GeoTIFF.' % filename)
+                print('The following errors were found:')
+                for error in errors:
+                    print(' - ' + error)
+                print('')
+            ret = 1
+        else:
+            if not quiet:
+                print('%s is a valid cloud optimized GeoTIFF' % filename)
+
+        if not quiet and not warnings and not errors:
+            headers_size = min(
+                details['data_offsets'][k] for k in details['data_offsets'])
+            if headers_size == 0:
+                headers_size = gdal.VSIStatL(filename).size
+            print('\nThe size of all IFD headers is %d bytes' % headers_size)
+    except ValidateCloudOptimizedGeoTIFFException as e:
+        if not quiet:
+            print('%s is NOT a valid cloud optimized GeoTIFF : %s' %
+                  (filename, str(e)))
+        ret = 1
+
+    return ret
+
+
+def save_as(filename, scratch_dir='.', flag_compress=True,
+            resamp_algorithm='average_magphase'):
```

Comment:
Thanks @vbrancat for this PR. All looks good. My only comment is the resamp_algorithm for BuildOverviews. I think by building overviews we are downsampling the dataset. Here we have complex data and the default resamp_algorithm is set to "average_magphase" which seems to me it will average the complex SLC data. But we can not really do that. We would if we had interferogram but not SLC. I think the best we can do when downsampling SLC is to decimate without any interpolation. So I think our only choices here would be NEAREST or NONE. Do you by any chance know what does NONE actually do in this case?

Reply:
It is very difficult find an explanation on what the actual resample algorithms do. In the GDAL guide (pp.193 https://gdal.org/gdal.pdf), `NONE` is not listed as a resample algorithm. My understanding is that when `NONE` is selected, no downsample is applied, which is what we wanted. What do you think? `NEAREST` seems to be a bit risky for complex data.